### ISSUE TITLE
Resource dependent tasks do not get re-distributed

### ISSUE THREAD

**Issue by MaxBenChrist** [dask/dask#3264]
Hi Dask Maintainers, thanks for this awesome package!

This issue contains the minimal example for behaviour described in the following stack overflow issue

https://stackoverflow.com/questions/48641051/redistribute-dask-tasks-among-the-cluster

So, I have a dask scheduler and a worker running.

```bash
dask-scheduler # is at 10.0.0.69:8786
dask-worker 10.0.0.69:8786 --resources foo=1
```

Now, I start a script that assigns, due to the foo resource requierement, one task to each worker at a time 

```python
from dask.distributed import Client
from time import sleep

if __name__ == "__main__":
    client = Client(address="10.0.0.69:8786")
    result = client.gather(client.map(sleep, [10] * 100, resources={'foo': 1}, pure=False))
```

If I have that script running and later add new dask workers, also with the same `dask-worker 10.0.0.69:8786 --resources foo=1` command, the dasks do not get redistributed among the cluster. The new workers stay unemployed.

<img width="972" alt="bildschirmfoto 2018-03-09 um 12 14 34" src="https://user-images.githubusercontent.com/13162426/37220142-a98d8f7a-2393-11e8-99a9-ba6bb1afac34.png">

What is interesting is, that if one of the workers dies, the tasks get still redistributed to the new workers. 

Is this expected behaviour? That resource dependent tasks do not get redistributed on the cluster if new workers are added?

Edit:

I use python 2.7 with the following packages

```bash
apipkg (1.4)
appnope (0.1.0)
attrs (17.3.0)
backports-abc (0.5)
backports.functools-lru-cache (1.4)
backports.shutil-get-terminal-size (1.0.0)
bleach (2.1.2)
bokeh (0.12.14)
certifi (2017.7.27.1)
chardet (3.0.4)
click (6.7)
cloudpickle (0.5.2)
configparser (3.5.0)
coverage (4.4.2)
cycler (0.10.0)
Cython (0.27.3)
dask (0.17.1)
dask-glm (0.1.0)
dask-ml (0.4.0)
dask-searchcv (0.2.0)
decorator (4.1.2)
distributed (1.21.3)
entrypoints (0.2.3)
enum34 (1.1.6)
execnet (1.5.0)
fastparquet (0.1.3)
funcsigs (1.0.2)
functools32 (3.2.3.post2)
future (0.16.0)
futures (3.2.0)
HeapDict (1.0.0)
html5lib (1.0.1)
idna (2.5)
imbalanced-learn (0.3.2)
ipaddress (1.0.19)
ipykernel (4.7.0)
ipython (5.3.0)
ipython-genutils (0.2.0)
Jinja2 (2.10)
jsonschema (2.6.0)
jupyter-client (5.2.0)
jupyter-core (4.4.0)
jupyterlab (0.31.8)
jupyterlab-launcher (0.10.5)
lightgbm (2.0.10)
llvmlite (0.21.0)
MarkupSafe (1.0)
matplotlib (2.1.1)
mistune (0.8.3)
mock (2.0.0)
msgpack-python (0.5.5)
multipledispatch (0.4.9)
nbconvert (5.3.1)
nbformat (4.4.0)
notebook (4.4.1)
numba (0.36.2)
numpy (1.14.1)
packaging (17.1)
pandas (0.22.0)
pandas-datareader (0.5.0)
pandoc (1.0.2)
pandocfilters (1.4.2)
pathlib2 (2.3.0)
patsy (0.4.1)
pbr (3.1.1)
pexpect (4.3.1)
pickleshare (0.7.4)
pip (9.0.1)
pkginfo (1.4.1)
pluggy (0.6.0)
ply (3.10)
prompt-toolkit (1.0.15)
protobuf (3.4.0)
psutil (5.4.3)
ptyprocess (0.5.2)
py (1.5.2)
pyarrow (0.7.1)
Pygments (2.2.0)
pyparsing (2.2.0)
pytest (3.3.1)
pytest-cov (2.5.1)
pytest-forked (0.2)
pytest-xdist (1.21.0)
python-dateutil (2.6.1)
pytz (2017.3)
PyYAML (3.12)
pyzmq (16.0.3)
requests (2.18.2)
requests-file (1.4.2)
requests-ftp (0.3.1)
requests-toolbelt (0.8.0)
scandir (1.6)
scikit-learn (0.19.1)
scipy (1.0.0)
seaborn (0.7.1)
setuptools (32.1.0)
simplegeneric (0.8.1)
singledispatch (3.4.0.3)
six (1.11.0)
sklearn-evaluation (0.4)
sortedcontainers (1.5.9)
statsmodels (0.8.0)
subprocess32 (3.2.7)
tblib (1.3.2)
terminado (0.8.1)
testpath (0.3.1)
thrift (0.11.0)
toolz (0.9.0)
tornado (5.0)
tqdm (4.19.6)
traitlets (4.3.2)
tsfresh (0.11.0)
twine (1.9.1)
urllib3 (1.22)
virtualenv (15.1.0)
wcwidth (0.1.7)
webencodings (0.5.1)
wheel (0.29.0)
zict (0.1.3)
```

**Comment by mrocklin** [comment]
Yup, work stealing doesn't currently look at tasks with extra constraints, like resources.  It would be nice to have this fixed.

**Comment by MaxBenChrist** [comment]
Thanks for the quick answer, the reason I add the `foo` resource is to make sure that each worker only processes one task at a time. Do you have an idea how to implement that behaviour without the  resource so I can take usage of the work stealing?

**Comment by mrocklin** [comment]
Resources are a great way of solving that problem, ideally someone would improve work stealing to support them.  

You might also consider a `dask.utils.SerializableLock`

**Comment by lesteve** [comment]
I think this can be closed since #3069 was merged.

**Comment by carusyte** [comment]
I stumbled into issues which might be related to task stealing and resource constraint.
Workers and tasks are specified with conceptual resource tag and numbers representing computing power supply and demand. 
Everything seems fine until a while some workers appear to have a long task backlog whereas others sitting idle. Work stealing seems not happening at all.
Is it true that work stealing won't take place if restrictions are present?
https://distributed.dask.org/en/latest/work-stealing.html#restrictions
